tcp_msg_ring.go

```diff
@@ -132,10 +132,16 @@ func (m *TCPMsgRing) disconnection(addr string) {
 	m.lock.Lock()
 	conn := m.conns[addr]
 	delete(m.conns, addr)
-	m.lock.Unlock()
 	if conn != nil {
-		conn.conn.Close()
+		if conn.conn != nil {
+			err := conn.conn.Close()
+			if err != nil {
+				log.Println("tcp msg ring disconnection close err:", err)
+			}
+		}
+
 	}
+	m.lock.Unlock()
```

Comment:
I'm curious why you moved this Unlock? Once the m.conns[addr] is cleared, shouldn't it be fine for somebody else to start a new connection to that addr while we play around with closing the last connection, logging it, etc.